File: packages/scrape-kdl/src/output-semantics.ts
import { ExecutionError } from "./execution-error.js";
import type { FieldIR, JsonValue } from "./ir.js";
import type { Warning } from "./public-api.js";

export class MissingValue extends Error {}

export interface OutputState {
  readonly warnings: Warning[];
  partial: boolean;
}

export function handleMissingOutput(field: FieldIR, path: string, message: string): JsonValue {
  if (field.required) throw new ExecutionError("E_REQUIRED_VALUE_MISSING", message, { path });
  return field.default ?? null;
}

export function recoverOutputField(state: OutputState, field: FieldIR, path: string, cause: unknown): JsonValue {
  if (field.onError === "fail") {
    if (cause instanceof ExecutionError) throw cause;
    throw new ExecutionError("E_FIELD_EXECUTION", errorMessage(cause), { path, cause });
  }
  state.partial = true;
  if (field.onError === "warn")
    state.warnings.push({ code: "W_ERROR_RECOVERED", message: executionMessage(cause), path });
  if (field.onError === "default") {
    if (field.default === undefined)
      throw new ExecutionError("E_IR_INVALID", "on-error default requires a field default", { path });
    return field.default;
  }
  return null;
}

export function validateCollectionMinimum(required: boolean, minItems: number, count: number, path: string): void {
  const minimum = required ? Math.max(1, minItems) : minItems;
  if (count < minimum)
    throw new ExecutionError(
      "E_COLLECTION_CARDINALITY",
      `collection has ${count} rows after recovery; minimum is ${minimum}`,
      { path },
    );
}

export function validateCollectionMaximum(maxItems: number | undefined, count: number, path: string): void {
  if (maxItems !== undefined && count > maxItems)
    throw new ExecutionError("E_COLLECTION_CARDINALITY", `collection has ${count} rows; maximum is ${maxItems}`, {
      path,
    });
}

export function executionMessage(error: unknown): string {
  return error instanceof ExecutionError && error.path !== undefined
    ? `${error.code} at ${error.path}: ${error.message}`
    : errorMessage(error);
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error ?? "operation canceled");
}

File: internal/executor/output_semantics.go
package executor

import (
	"fmt"

	"github.com/hsblabs/scrape-kdl/internal/ir"
)

func handleMissingOutput(field ir.Field, path, message string) (any, error) {
	if field.Required {
		return nil, &ExecutionError{Code: "E_REQUIRED_VALUE_MISSING", Message: message, Path: path}
	}
	if field.Default != nil {
		return decodeFieldDefault(field, path)
	}
	return nil, nil
}

func recoverOutputField(partial *bool, warnings *[]Warning, field ir.Field, path string, cause error) (any, error) {
	policy := field.OnError
	if policy == "" {
		if field.Required {
			policy = "fail"
		} else {
			policy = "null"
		}
	}
	switch policy {
	case "fail":
		if execution, ok := cause.(*ExecutionError); ok {
			return nil, execution
		}
		return nil, &ExecutionError{Code: "E_FIELD_EXECUTION", Message: cause.Error(), Path: path, Cause: cause}
	case "null":
		*partial = true
		return nil, nil
	case "warn":
		*partial = true
		*warnings = append(*warnings, Warning{Code: "W_ERROR_RECOVERED", Message: cause.Error(), Path: path})
		return nil, nil
	case "default":
		value, err := decodeFieldDefault(field, path)
		if err != nil {
			return nil, err
		}
		*partial = true
		return value, nil
	default:
		return nil, &ExecutionError{Code: "E_IR_INVALID", Message: fmt.Sprintf("unknown on-error policy %q", policy), Path: path}
	}
}

func validateCollectionMinimum(required bool, minItems, count int, path string) error {
	minimum := minItems
	if required && minimum < 1 {
		minimum = 1
	}
	if count < minimum {
		return &ExecutionError{Code: "E_COLLECTION_CARDINALITY", Message: fmt.Sprintf("collection has %d rows after recovery; minimum is %d", count, minimum), Path: path}
	}
	return nil
}

func validateCollectionMaximum(maxItems *int, count int, path string) error {
	if maxItems != nil && count > *maxItems {
		return &ExecutionError{Code: "E_COLLECTION_CARDINALITY", Message: fmt.Sprintf("collection has %d rows; maximum is %d", count, *maxItems), Path: path}
	}
	return nil
}
